src/Features/CSharp/Portable/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,171 @@
+{
+    using static Helpers;
+
+    /// <summary>
+    /// Analyzer that looks for several variants of code like `s.Slice(start, end - start)` and
+    /// offers to update to `s[start..end]` or `s.Slice(start..end)`.  In order to convert to the
+    /// indexer, the type being called on needs a slice-like method that takes two ints, and returns
+    /// an instance of the same type. It also needs a Length/Count property, as well as an indexer
+    /// that takes a System.Range instance.  In order to convert between methods, there need to be
+    /// two overloads that are equivalent except that one takes two ints, and the other takes a
+    /// System.Range.
+    ///
+    /// It is assumed that if the type follows this shape that it is well behaved and that this
+    /// transformation will preserve semantics.  If this assumption is not good in practice, we
+    /// could always limit the feature to only work on a whitelist of known safe types.
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp), Shared]
+    internal partial class CSharpUseRangeOperatorDiagnosticAnalyzer : AbstractCodeStyleDiagnosticAnalyzer
+    {
+        public const string UseIndexer = nameof(UseIndexer);
+        public const string ComputedRange = nameof(ComputedRange);
+        public const string ConstantRange = nameof(ConstantRange);
+
+        public CSharpUseRangeOperatorDiagnosticAnalyzer() 
+            : base(IDEDiagnosticIds.UseRangeOperatorDiagnosticId,
+                   new LocalizableResourceString(nameof(FeaturesResources.Use_range_operator), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                   new LocalizableResourceString(nameof(FeaturesResources._0_can_be_simplified), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+        }
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(compilationContext =>
+            {
+                // We're going to be checking every invocation in the compilation. Cache information
+                // we compute in this object so we don't have to continually recompute it.
+                var infoCache = new InfoCache(compilationContext.Compilation);
+                compilationContext.RegisterOperationAction(
+                    c => AnalyzeInvocation(c, infoCache),
+                    OperationKind.Invocation);
+            });
+        }
+
+        private void AnalyzeInvocation(
+            OperationAnalysisContext context, InfoCache infoCache)
+        {
+            var cancellationToken = context.CancellationToken;
+            var invocation = (IInvocationOperation)context.Operation;
+
+            var invocationSyntax = invocation.Syntax as InvocationExpressionSyntax;
+            if (invocationSyntax is null ||
+                invocationSyntax.ArgumentList is null)
+            {
+                return;
+            }
```

Comment:
Why do we need the syntax check? Probably you can bail out on `invocation.IsImplicitDeclared` if you are worried about compiler generated invocations.

Reply:
the syntax part is important because the fixer presumes a certain shape to the syntax to do its rewriting.  so we want to at lteast verify that we're on a piece of syntax that is sensible to us.

Reply:
i'll comment accordingly.